extension/clientaddrmiddlewareextension/middleware.go

```diff
@@ -34,26 +38,78 @@ func newClientAddrMiddleware(cfg *Config, set extension.Settings) (*clientAddrMi
 	return &clientAddrMiddleware{cfg: cfg}, nil
 }
 
+// Start starts the middleware extension.
 func (c *clientAddrMiddleware) Start(ctx context.Context, host component.Host) error {
 	return nil
 }
 
+// Shutdown shuts down the middleware extension.
 func (c *clientAddrMiddleware) Shutdown(ctx context.Context) error {
 	return nil
 }
 
+// ctxWithClientAddr returns a new context with updated client info that contains
+// a valid ip address found in the headers.
+// If no client address is found, the original context is returned.
+func ctxWithClientAddr(ctx context.Context, headers map[string][]string) context.Context {
+	if ip := netutil.ClientAddrFromHeaders(headers); ip != nil {
+		cl := client.FromContext(ctx)
+		cl.Addr = ip
+		return client.NewContext(ctx, cl)
+	}
+	return ctx
+}
+
+// GetHTTPHandler HTTP server middleware that returns an HTTP handler that updates the ctx client.Info.Address
+// using request headers.
 func (c *clientAddrMiddleware) GetHTTPHandler(base http.Handler) (http.Handler, error) {
-	return base, nil
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := ctxWithClientAddr(r.Context(), r.Header)
+		base.ServeHTTP(w, r.WithContext(ctx))
+	}), nil
 }
 
+// GetGRPCServerOptions GRPC server middleware that returns gRPC server options that updates the ctx client.Info.Address
+// using request metadata.
 func (c *clientAddrMiddleware) GetGRPCServerOptions() ([]grpc.ServerOption, error) {
-	return []grpc.ServerOption{}, nil
+	opt := grpc.ChainUnaryInterceptor(
+		func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+			md, ok := metadata.FromIncomingContext(ctx)
+			if ok {
+				ctx = ctxWithClientAddr(ctx, md)
+			}
+			return handler(ctx, req)
+		},
+	)
+	return []grpc.ServerOption{opt}, nil
+}
+
+type clientAddrRoundTripper struct {
+	base http.RoundTripper
+}
+
+func (c *clientAddrRoundTripper) RoundTrip(r *http.Request) (*http.Response, error) {
+	ctx := ctxWithClientAddr(r.Context(), r.Header)
+	return c.base.RoundTrip(r.WithContext(ctx))
 }
 
-func (c *clientAddrMiddleware) GetHTTPRoundTripper(tripper http.RoundTripper) (http.RoundTripper, error) {
-	return tripper, nil
+// GetHTTPRoundTripper HTTP client middleware that returns a HTTP round tripper that updates the ctx client.Info.Address
```

Comment:
Sorry, I overlooked this before: I don't think we need GetHTTPRoundTripper or GetGRPCClientOptions. These are for HTTP/gRPC client requests. We only need to enrich the context for server requests,right?

Reply:
Yeah let's remove the client middleware